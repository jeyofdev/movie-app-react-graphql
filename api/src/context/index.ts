import PersonService from '../services/Person.service';
import CollectionService from '../services/Collection.service';
import MovieService from '../services/Movie.service';
import KeywordService from '../services/Keyword.service';
import GenreService from '../services/Genre.service';
import SearchService from '../services/Search.service';

export interface IContext {
	dataSource: {
		movies: MovieService;
		collections: CollectionService;
		persons: PersonService;
		keywords: KeywordService;
		genres: GenreService;
		search: SearchService;
	};
}

export const context: IContext = {
	dataSource: {
		movies: new MovieService(),
		collections: new CollectionService(),
		persons: new PersonService(),
		keywords: new KeywordService(),
		genres: new GenreService(),
		search: new SearchService(),
	},
};
