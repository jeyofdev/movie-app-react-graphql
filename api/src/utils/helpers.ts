import request from 'supertest';
import { OptionsInput } from '../__generated__/resolvers-types';

export const formatUrlQuery = (
	baseUrl: string,
	apiKey: string,
	pathParams?: string | number,
	queryString?: OptionsInput,
) => {
	let url = `${baseUrl}/`;
	url += pathParams ?? `${pathParams}`;
	url += `?api_key=${apiKey}&`;

	if (queryString) {
		const allQueryStrings = Object.entries(queryString)
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			.filter(([_, value]) => value)
			.map(value => value.join('='));

		url += allQueryStrings.join('&');
	}
	return url;
};

export const executeRequestForTesting = async (
	url: string,
	query: string,
	variables: unknown,
	statusCode = 200,
) => {
	const response = await request(url)
		.post('/')
		.set('Accept', 'application/json')
		.send({
			query,
			variables,
		})
		.expect(statusCode);

	return response;
};
