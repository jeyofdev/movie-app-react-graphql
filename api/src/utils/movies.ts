import { MoviesResponse } from '../__generated__/resolvers-types';

const setMoviesResult = (movies: MoviesResponse) => ({
	page: movies.page,
	results: movies.results,
	total_pages: movies.total_pages,
	total_results: movies.total_results,
});

export default setMoviesResult;
