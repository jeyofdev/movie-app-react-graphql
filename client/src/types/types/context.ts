import { LanguageEnum } from '@graphql/__generated__/graphql-type';
import { DarkModeEnum } from '../enums';

export type ThemeContextType = {
	themeMode: DarkModeEnum;
	handleThemeMode: () => void;
};

export type TranslationContextType = {
	currentLocale: LanguageEnum;
	handleLocale: (locale: LanguageEnum) => void;
};
