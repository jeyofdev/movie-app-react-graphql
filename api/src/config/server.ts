import { startStandaloneServer } from '@apollo/server/standalone';
import { readFileSync } from 'fs';
import { ApolloServer } from '@apollo/server';
import { IContext, context } from '../context';
import resolvers from '../resolvers';

const getServer = async (port?: number) => {
	const typeDefs = readFileSync('./src/schema/schema.graphql', {
		encoding: 'utf-8',
	});

	const server = new ApolloServer<IContext>({
		typeDefs,
		resolvers,
	});

	const { url } = await startStandaloneServer(server, {
		context: async () => context,
		listen: { port },
	});

	return { server, url };
};

export default getServer;
