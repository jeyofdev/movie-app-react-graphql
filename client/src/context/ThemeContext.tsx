import { createContext, useState } from 'react';
import { DarkModeEnum } from '../types/enums';
import { ThemeContextType } from '../types/types/context';
import { ThemeContextProviderPropsType } from '../types/types/props';

export const ThemeContext = createContext<ThemeContextType>({
	themeMode: DarkModeEnum.DARK,
	handleThemeMode: () => {},
});

const ThemeContextProvider = ({ children }: ThemeContextProviderPropsType) => {
	const [themeMode, setThemeMode] = useState<DarkModeEnum>(DarkModeEnum.DARK);

	const handleThemeMode = () => {
		setThemeMode(
			themeMode === DarkModeEnum.DARK ? DarkModeEnum.LIGHT : DarkModeEnum.DARK,
		);
	};

	return (
		<ThemeContext.Provider value={{ themeMode, handleThemeMode }}>
			{children}
		</ThemeContext.Provider>
	);
};

export default ThemeContextProvider;
