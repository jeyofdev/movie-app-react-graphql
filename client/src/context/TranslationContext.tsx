import { LanguageEnum } from '@graphql/__generated__/graphql-type';
import { createContext, useState } from 'react';
import { TranslationContextType } from '../types/types/context';
import { ThemeContextProviderPropsType } from '../types/types/props';

export const TranslationContext = createContext<TranslationContextType>({
	currentLocale: LanguageEnum.En,
	handleLocale: () => {},
});

const TranslationContextProvider = ({
	children,
}: ThemeContextProviderPropsType) => {
	const [currentLocale, setcurrentLocale] = useState<LanguageEnum>(
		(localStorage.getItem('locale') as LanguageEnum) ?? LanguageEnum.En,
	);

	const handleLocale = (locale: LanguageEnum) => {
		setcurrentLocale(locale);
	};

	return (
		<TranslationContext.Provider value={{ currentLocale, handleLocale }}>
			{children}
		</TranslationContext.Provider>
	);
};

export default TranslationContextProvider;
