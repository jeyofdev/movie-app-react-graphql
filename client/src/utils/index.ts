import { format } from 'date-fns';
import { enGB, fr, es, de } from 'date-fns/locale';
import { toast } from 'react-toastify';
import { LanguageEnum } from '@graphql/__generated__/graphql-type';
import { RoutesEnum } from '../types/enums';

/**
 * Limit text by number words
 * @param text
 * @param limit
 * @returns
 */
export const truncate = (text: string, limit: number): string => {
	const { length } = text.split(' ');

	return length <= limit
		? text
		: `${text.split(' ').slice(0, limit).join(' ')}...`;
};

/**
 * Format number to hours and minutes
 * @param totalMinutes
 * @returns
 */
export const formatNumberToHours = (totalMinutes: number): string => {
	const hours = Math.floor(totalMinutes / 60);
	const minutes = totalMinutes % 60;

	if (minutes === 0) {
		return `${hours}h`;
	}

	if (minutes < 10) {
		return `${hours}h ${`0${minutes}`}min`;
	}

	return `${hours}h ${minutes}min`;
};

export const formatGenreForUrl = (genre: string) =>
	`/${RoutesEnum.MOVIES_GENRE}/${genre.toLowerCase().split(' ').join('-')}`;

export const firstLetterCapitalize = (text: string): string =>
	text.slice(0, 1).toUpperCase() + text.slice(1);

export const getAge = (birthday: string) => {
	const date = new Date(birthday);
	const diff = Date.now() - date.getTime();
	const age = new Date(diff);

	return Math.abs(age.getUTCFullYear() - 1970);
};

export const getAgeBetweenTwoDate = (birthday: string, deathday: string) => {
	const dateBirthday = new Date(birthday);
	const dateDeathday = new Date(deathday);
	const diff = dateDeathday.getTime() - dateBirthday.getTime();
	const age = new Date(diff);

	return Math.abs(age.getUTCFullYear() - 1970);
};

export const formatDate = (date: string, locale: LanguageEnum) => {
	const newDate = new Date(date);

	const formatByLocale = () => {
		switch (locale) {
			case LanguageEnum.En:
				return 'MMMM dd, yyyy';
			case LanguageEnum.Fr:
				return 'dd MMMM yyyy';
			case LanguageEnum.Es:
				return 'dd MMMM yyy';
			case LanguageEnum.De:
				return 'dd. MMMM yyyy';

			default:
				return 'MMMM dd, yyyy';
		}
	};

	const getLocale = () => {
		switch (locale) {
			case LanguageEnum.En:
				return enGB;
			case LanguageEnum.Fr:
				return fr;
			case LanguageEnum.Es:
				return es;
			case LanguageEnum.De:
				return de;

			default:
				return enGB;
		}
	};

	return format(newDate, formatByLocale(), { locale: getLocale() });
};

export const toastSuccess = (message: string, autoClose?: number) =>
	toast.success(message, {
		position: 'top-right',
		autoClose: autoClose ?? 5000,
		hideProgressBar: false,
		closeOnClick: true,
		pauseOnHover: true,
		draggable: true,
		progress: undefined,
		theme: 'colored',
	});
