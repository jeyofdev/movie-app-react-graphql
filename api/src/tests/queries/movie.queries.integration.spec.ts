import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { ApolloServer } from '@apollo/server';
import { IContext } from '../../context';
import getServer from '../../config/server';
import 'dotenv/config';
import { executeRequestForTesting } from '../../utils/helpers';
import moviesQueryOperations from '../../operations/movie.query.operations';

describe('init server', () => {
	let server: ApolloServer<IContext>;
	let url: string;

	beforeAll(async () => {
		if (!process.env.TEST_API_PORT) {
			throw new Error(
				'The environment variable TEST_API_PORT must be specified',
			);
		}
		({ server, url } = await getServer());
	});

	afterAll(async () => {
		await server.stop();
	});

	describe('queries movies', () => {
		it('When no movie found', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.details,
				{
					movieId: 7660450,
					options: {
						language: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.movieDetails).toBeNull();
		});

		it('Get the primary information about a movie.', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.details,
				{
					movieId: 76600,
					options: {
						language: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.movieDetails).toMatchSnapshot();
		});

		it('Returns a list of the current popular movies', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.popular,
				{
					options: {
						language: 'FR',
						page: 1,
						region: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.popularMovies).toMatchSnapshot();
		});

		it('Returns a list of movies in theatres', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.nowPlaying,
				{
					options: {
						language: 'FR',
						page: 1,
						region: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.nowPlayingMovies).toMatchSnapshot();
		});

		it('Returns a list of upcoming movies in theatres', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.upcoming,
				{
					options: {
						language: 'FR',
						page: 1,
						region: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.upcomingMovies).toMatchSnapshot();
		});

		it('Returns the top rated movies on TMDB', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.topRating,
				{
					options: {
						language: 'FR',
						page: 1,
						region: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.topRatedMovies).toMatchSnapshot();
		});

		it('Returns the cast for a movie', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.cast,
				{
					movieId: 76600,
					options: {
						language: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.castMovie).toMatchSnapshot();
		});

		it('Returns the crew for a movie', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.crew,
				{
					movieId: 76600,
					options: {
						language: 'FR',
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.crewMovie).toMatchSnapshot();
		});

		it('Returns a list of similar movies', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.similar,
				{
					movieId: 76600,
					options: {
						language: 'ES',
						page: 1,
					},
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.similarMovie).toMatchSnapshot();
		});

		it('Returns the images that belong to a movie', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.images,
				{
					movieId: 76600,
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.imagesMovie).toMatchSnapshot();
		});

		it('Returns the keywords that have been added to a movie', async () => {
			const response = await executeRequestForTesting(
				url,
				moviesQueryOperations.keywords,
				{
					movieId: 25,
				},
			);

			expect(response.error).not.toBeUndefined();
			expect(response.body.data.keywordsMovie).toMatchSnapshot();
		});
	});
});
