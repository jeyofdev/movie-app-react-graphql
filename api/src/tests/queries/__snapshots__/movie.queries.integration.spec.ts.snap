// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`init server > query MovieDetails > when a movie matches 1`] = `
{
  "adult": false,
  "backdrop_path": "/8rpDcsfLJypbO6vREc0547VKqEv.jpg",
  "budget": 460000000,
  "genre_ids": null,
  "genres": [
    {
      "id": 878,
      "name": "Science Fiction",
    },
    {
      "id": 12,
      "name": "Adventure",
    },
    {
      "id": 28,
      "name": "Action",
    },
  ],
  "homepage": "https://www.avatar.com/movies/avatar-the-way-of-water",
  "id": 76600,
  "imdb_id": "tt1630029",
  "original_language": "en",
  "original_title": "Avatar: The Way of Water",
  "overview": "Set more than a decade after the events of the first film, learn the story of the Sully family (Jake, Neytiri, and their kids), the trouble that follows them, the lengths they go to keep each other safe, the battles they fight to stay alive, and the tragedies they endure.",
  "popularity": 2857.869,
  "poster_path": "/t6HIqrRAclMCA60NsSmeqe9RmNV.jpg",
  "production_companies": [
    {
      "id": 127928,
      "logo_path": "/h0rjX5vjW5r8yEnUBStFarjcLT4.png",
      "name": "20th Century Studios",
      "origin_country": "US",
    },
    {
      "id": 574,
      "logo_path": "/iB6GjNVHs5hOqcEYt2rcjBqIjki.png",
      "name": "Lightstorm Entertainment",
      "origin_country": "US",
    },
  ],
  "production_countries": [
    {
      "iso_3166_1": "US",
      "name": "United States of America",
    },
  ],
  "release_date": "2022-12-14",
  "runtime": 192,
  "spoken_languages": [
    {
      "iso_639_1": "en",
      "name": "English",
    },
  ],
  "status": "Released",
  "tagline": "Return to Pandora.",
  "title": "Avatar: The Way of Water",
  "video": false,
  "vote_average": 7.707,
  "vote_count": 7832,
}
`;
